engine/engine/content/builtins/render/default.render_script

```diff
@@ -3,92 +3,177 @@
 end
 --
 -- projection that stretches content
 --
-local function stretch_projection(near, far)
-    return vmath.matrix4_orthographic(0, render.get_width(), 0, render.get_height(), near, far)
+local function get_stretch_projection(camera, state)
+    return vmath.matrix4_orthographic(0, state.width, 0, state.height, camera.near, camera.far)
+end
+--
+-- projection for gui
+--
+local function get_gui_projection(camera, state)
+    return vmath.matrix4_orthographic(0, state.window_width, 0, state.window_height, camera.near, camera.far)
 end
 
-local function get_projection(self)
-    return self.projection_fn(self.near, self.far, self.zoom)
+local function update_clear_color(state, color)
+    if color then
+        state.clear_buffers[render.BUFFER_COLOR_BIT] = color
+    end
 end
 
-function init(self)
-    self.tile_pred = render.predicate({"tile"})
-    self.gui_pred = render.predicate({"gui"})
-    self.text_pred = render.predicate({"text"})
-    self.particle_pred = render.predicate({"particle"})
+local function update_camera(camera, state)
+    camera.proj = camera.projection_fn(camera, state)
+    camera.frustum.frustrum = camera.proj * camera.view
+end
+
+local function update_state(state)
+    state.window_width = render.get_window_width()
+    state.window_height = render.get_window_height()
+    state.valid = state.window_width > 0 and state.window_height > 0
+    if not state.valid then
+        return false
+    end
+    -- Make sure state updated only once when resize window
+    if state.window_width == state.prev_window_width and state.window_height == state.prev_window_height then
+        return true
+    end
+    state.prev_window_width = state.window_width
+    state.prev_window_height = state.window_height
+    state.width = render.get_width()
+    state.height = render.get_height()
+    for _, camera in pairs(state.cameras) do
+        update_camera(camera, state)
+    end
+    return true
+end
+
+local function init_camera(camera, projection_fn, near, far, zoom)
+    camera.view = vmath.matrix4()
+    camera.near = near == nil and DEFAULT_NEAR or near
+    camera.far = far == nil and DEFAULT_FAR or far
+    camera.zoom = zoom == nil and DEFAULT_ZOOM or zoom
+    camera.projection_fn = projection_fn
+end
 
-    self.clear_color = vmath.vector4(0, 0, 0, 0)
-    self.clear_color.x = sys.get_config("render.clear_color_red", 0)
-    self.clear_color.y = sys.get_config("render.clear_color_green", 0)
-    self.clear_color.z = sys.get_config("render.clear_color_blue", 0)
-    self.clear_color.w = sys.get_config("render.clear_color_alpha", 0)
+local function create_predicates(...)
+    local arg = {...}
+    local predicates = {}
+    for _, predicate_name in pairs(arg) do
+        predicates[predicate_name] = render.predicate({predicate_name})
+    end
+    return predicates
+end
 
-    self.view = vmath.matrix4()
+local function create_camera(state, name, is_main_camera)
+    local camera = {}
+    camera.frustum = {}
```

Comment:
What is the point of having camera.frustum.frustum and not only camera.frustum? Are you thinking that we might want to store other options/settings for the frustum in camera.frustum table?

Reply:
we should send a table {frustum = ...}
so, ofc we can send the whole `camera` with many other fields, but I would prefer to have it isolated

Reply:
Ah, yes, ofc, so that we create the table only once. Makes sense!